### PR TITLE
PBR shader support Iridescence

### PR DESCRIPTION
**Pull request by hhhhkrx**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## Summary by CodeRabbit

- **New Features**
	- Introduced support for iridescence effects in materials, enhancing visual rendering capabilities.
	- Added new properties and interfaces for iridescence, allowing for more complex material definitions in 3D models.
	- Implemented a new GLTF extension for managing iridescence in materials.

- **Bug Fixes**
	- Updated existing interfaces to allow optional properties for greater flexibility in material construction.

- **Documentation**
	- Enhanced material parsing capabilities in the GLTF loader to include iridescence properties.

### CODE REVIEW

File: packages/core/src/material/PBRMaterial.ts [hunk 1]
```diff
@@ -132,6 +136,98 @@ export class PBRMaterial extends PBRBaseMaterial {
     }
   }
 
+  /**
+   * The iridescence intensity factor.	
+   * @defaultValue `1.0`
```
Comment (zhuxudong) [inline packages/core/src/material/PBRMaterial.ts:141]: default is 0

File: packages/loader/src/gltf/extensions/GLTFExtensionSchema.ts [hunk 2]
```diff
@@ -26,11 +26,11 @@ export interface IKHRLightsPunctual {
...
-  clearcoatFactor: number;
-  clearcoatTexture: ITextureInfo;
-  clearcoatRoughnessFactor: number;
-  clearcoatRoughnessTexture: ITextureInfo;
-  clearcoatNormalTexture: IMaterialNormalTextureInfo;
+  clearcoatFactor?: number;
+  clearcoatTexture?: ITextureInfo;
+  clearcoatRoughnessFactor?: number;
+  clearcoatRoughnessTexture?: ITextureInfo;
+  clearcoatNormalTexture?: IMaterialNormalTextureInfo;
```
Comment (zhuxudong) [inline packages/loader/src/gltf/extensions/GLTFExtensionSchema.ts:33]: thank you

File: packages/loader/src/gltf/extensions/KHR_materials_iridescence.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      iridescenceIor = 1.3,
+      iridescenceThicknessMinimum = 100,
+      iridescenceThicknessMaximum = 400,
+      iridescenceThicknessTexture
+    } = schema;
+
+     material.iridescenceFactor = iridescenceFactor;
+     material.iridescenceIor = iridescenceIor;
+     material.iridescenceThicknessMin = iridescenceThicknessMaximum;
+     material.iridescenceThicknessMax = iridescenceThicknessMinimum;
```
Comment (zhuxudong) [inline packages/loader/src/gltf/extensions/KHR_materials_iridescence.ts:23]: revert min/max

File: packages/core/src/material/PBRMaterial.ts [hunk 4]
```diff
@@ -132,6 +136,98 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+
+  set iridescenceThicknessMax(value: number) {
+    const IridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceProp);
+    IridescenceInfo.w = value;
+  }
+  
+  /**
+   * The thickness texture of the thin-film layer.	
+   */
+  get irithicknessTexture(): Texture2D {
```
Comment (zhuxudong) [inline packages/core/src/material/PBRMaterial.ts:200]: `iri` is not a good abbreviation, use `iridescenceThicknessTexture` to instead

File: packages/core/src/material/PBRMaterial.ts [hunk 5]
```diff
@@ -132,6 +136,98 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+  /**
+   * The maximum thickness of the thin-film layer given in nanometers.	
+   * @defaultValue `400`
+   */
+  get iridescenceThicknessMax(): number {
+    return this.shaderData.getVector4(PBRMaterial._iridescenceProp).w;
+  }
+
+  set iridescenceThicknessMax(value: number) {
+    const IridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceProp);
```
Comment (zhuxudong) [inline packages/core/src/material/PBRMaterial.ts:193]: variable start with lowercase, such as `iridescenceInfo`

File: packages/core/src/material/PBRMaterial.ts [hunk 6]
```diff
@@ -132,6 +136,98 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+  /**
+   * The index of refraction of the dielectric thin-film layer.	
+   * @defaultValue `1.3`
+   */
+  get iridescenceIor(): number {
+    return this.shaderData.getVector4(PBRMaterial._iridescenceProp).y;
+  }
+
+  set iridescenceIor(value: number) {
+    const IridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceProp);
```
Comment (zhuxudong) [inline packages/core/src/material/PBRMaterial.ts:167]: the same as below

**Comment by zhuxudong** [comment]
do not submit code bypass the hook such as`prettier`, it cause some ci and symbol error here

File: packages/core/src/material/PBRMaterial.ts [hunk 7]
```diff
@@ -132,6 +136,99 @@ export class PBRMaterial extends PBRBaseMaterial {
     }
   }
 
+  /**
+   * The iridescence intensity factor.
+   * @defaultValue `0.0`
+   */
+  get iridescenceFactor(): number {
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:143]: It is better to use iridescenceIntensity uniformly

**Review by GuoLei1990 — APPROVED** [review]

File: packages/core/src/material/PBRMaterial.ts [hunk 8]
```diff
@@ -132,6 +136,99 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+   * @defaultValue `1.3`
+   */
+  get iridescenceIor(): number {
+    return this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp).y;
+  }
+
+  set iridescenceIor(value: number) {
+    const iridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp);
+    iridescenceInfo.y = value;
+  }
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:170]:
```ts
  set iridescenceIor(value: number) {
    const iridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp);
    iridescenceInfo.y = Math.max(value, 0);
  }
```

Ior can less than 0?

File: packages/core/src/material/PBRMaterial.ts [hunk 9]
```diff
@@ -132,6 +136,99 @@
...
+    const iridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp);
+    if (!!iridescenceInfo.x !== !!value) {
+      if (value === 0) {
+        this.shaderData.disableMacro("MATERIAL_ENABLE_IRIDESCENCE");
+      } else {
+        this.shaderData.enableMacro("MATERIAL_ENABLE_IRIDESCENCE");
+      }
+      iridescenceInfo.x = value;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/material/PBRMaterial.ts]: +1

File: packages/core/src/material/PBRMaterial.ts [hunk 10]
```diff
@@ -132,6 +136,99 @@ export class PBRMaterial extends PBRBaseMaterial {
...
 
+  /**
+   * The iridescence intensity factor.
+   * @defaultValue `0.0`
+   */
+  get iridescence(): number {
+    return this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp).x;
+  }
+
+  set iridescence(value: number) {
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:147]:
```ts
set iridescence(value: number) {
    value = Math.max(0, Math.min(1, value));
    const iridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp);
    if (!!iridescenceInfo.x !== !!value) {
      if (value === 0) {
        this.shaderData.disableMacro("MATERIAL_ENABLE_IRIDESCENCE");
      } else {
        this.shaderData.enableMacro("MATERIAL_ENABLE_IRIDESCENCE");
      }
    }
    iridescenceInfo.x = value;
  }
```

File: packages/core/src/material/PBRMaterial.ts [hunk 11]
```diff
@@ -132,6 +136,100 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+   * @defaultValue `400`
+   */
+  get iridescenceThicknessMax(): number {
+    return this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp).w;
+  }
+
+  set iridescenceThicknessMax(value: number) {
+    const iridescenceInfo = this.shaderData.getVector4(PBRMaterial._iridescenceInfoProp);
+    iridescenceInfo.w = value;
+  }
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:197]: Merge min and max ad `iridescenceThicknessRange` is better!

File: packages/core/src/material/PBRMaterial.ts [hunk 12]
```diff
@@ -132,6 +136,100 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+
+  /**
+   * The thickness texture of the thin-film layer.
+   */
+  get iridescenceThicknessTexture(): Texture2D {
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:202]: The comments are too simple, at least explain which channel is used

File: packages/core/src/material/PBRMaterial.ts [hunk 13]
```diff
@@ -132,6 +136,100 @@ export class PBRMaterial extends PBRBaseMaterial {
...
+      this.shaderData.enableMacro("MATERIAL_HAS_IRIDESCENCE_THICKNESS_TEXTURE");
+    } else {
+      this.shaderData.disableMacro("MATERIAL_HAS_IRIDESCENCE_THICKNESS_TEXTURE");
+    }
+  }
+
+  /**
+   * The iridescence intensity texture.
+   */
+  get iridescenceTexture(): Texture2D {
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRMaterial.ts:219]: The same suggestion with thickness texture

**Pull request merged**